# 1. 데이터 필터

1. kofic 일별 박스오피스 데이터 (최종_박스오피스_데이터.csv)
2. cgv 줄거리 파일 (cgv_장르_줄거리_통합본.csv) - 기준
분석에 사용할 박스오피스 데이터만 걸러낸다

In [728]:
import pandas as pd
import numpy as np

In [729]:
kofic_data=pd.read_csv("../영화관람객크롤링/최종_박스오피스_데이터.csv",encoding='utf-8-sig')
cgv_data=pd.read_csv("../영화리뷰크롤링/cgv리뷰_전처리/전처리파일저장/cgv_장르_줄거리_통합본.csv",encoding='utf-8-sig')

In [730]:
kofic_data.head(1)

영화명  개봉일    매출액  관객수  스크린수  상영횟수 대표국적  배급사   등급  장르   감독  \
0  #Take No. 고백(인디스데이)  NaN  13000    2     1     1   한국  NaN  NaN  기타  NaN   

    배우          일자  개봉 구분  
0  NaN  2023-02-10  최초 개봉

In [731]:
cgv_data.head(1)

영화제목     장르                                                줄거리
0  좋은날  애니메이션  20대의 풋풋한 사랑 [봄봄], 40대의 처참했던 슬픔 [운수 좋은 날] 그리고 6...

데이터 중복 확인

In [732]:
kofic_data[["영화명","일자"]].duplicated().sum()

np.int64(1894)

중복 데이터 1894개 제거

In [733]:
kofic_data.drop_duplicates(subset=["영화명","일자"],inplace=True)

# 2. 사용할 영화

In [734]:
barometer=cgv_data["영화제목"]
temp=kofic_data[kofic_data["영화명"].isin(barometer)]

In [735]:
temp["영화명"].nunique()

6472

# 3. 최초개봉만

In [736]:
temp=temp[temp["개봉 구분"]=="최초 개봉"]
temp.head()

영화명         개봉일      매출액   관객수  스크린수  상영횟수 대표국적                배급사  \
12  #살아있다  2020-06-24   994000   142     1     1   한국  롯데컬처웍스(주)롯데엔터테인먼트   
13  #살아있다  2020-06-24   909000   449     2     2   한국  롯데컬처웍스(주)롯데엔터테인먼트   
14  #살아있다  2020-06-24  7314000  1386     7    11   한국  롯데컬처웍스(주)롯데엔터테인먼트   
15  #살아있다  2020-06-24  3950000   790     6     6   한국  롯데컬처웍스(주)롯데엔터테인먼트   
16  #살아있다  2020-06-24  2503000  1239     5     5   한국  롯데컬처웍스(주)롯데엔터테인먼트   

          등급   장르   감독                                                 배우  \
12  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
13  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
14  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
15  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
16  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   

            일자  개봉 구분  
12  2020-06-02  최초 개봉  
13  2020-06-03  최초 개봉  
14  2020-06-15  최초 개봉  
15  2020-06-19  최초 개봉  
16  2020-06-23  최초 개봉

# 4. n주차 생성

개봉일이 없는 영화 - 그 영화는 모든 일자에 개봉일이 없나?  
개봉일이 하나도 없는 영화는 제거

In [737]:
temp = temp[temp.groupby('영화명')['개봉일'].transform('count') > 0]

In [738]:
temp.isna().sum()

영화명          0
개봉일         86
매출액          0
관객수          0
스크린수         0
         ...  
장르         143
감독        7709
배우       16323
일자           0
개봉 구분        0
Length: 14, dtype: int64

개봉일이 있는것과 없는 영화도 있음  
결측치 보간

In [739]:
temp['개봉일'] = temp.groupby('영화명')['개봉일'].transform('first')

In [740]:
temp.isna().sum()

영화명          0
개봉일          0
매출액          0
관객수          0
스크린수         0
         ...  
장르         143
감독        7709
배우       16323
일자           0
개봉 구분        0
Length: 14, dtype: int64

날짜 형식 변경  
  
1주차: 영화 개봉일이 포함된 주의 일요일까지  
2주차 이후: 월요일부터 일요일까지를 한 주로 계산

In [741]:
temp['개봉일_dt'] = pd.to_datetime(temp['개봉일'], errors='coerce')
temp['일자_dt'] = pd.to_datetime(temp['일자'], errors='coerce')

In [742]:
temp.head()

영화명         개봉일      매출액   관객수  스크린수  상영횟수 대표국적                배급사  \
12  #살아있다  2020-06-24   994000   142     1     1   한국  롯데컬처웍스(주)롯데엔터테인먼트   
13  #살아있다  2020-06-24   909000   449     2     2   한국  롯데컬처웍스(주)롯데엔터테인먼트   
14  #살아있다  2020-06-24  7314000  1386     7    11   한국  롯데컬처웍스(주)롯데엔터테인먼트   
15  #살아있다  2020-06-24  3950000   790     6     6   한국  롯데컬처웍스(주)롯데엔터테인먼트   
16  #살아있다  2020-06-24  2503000  1239     5     5   한국  롯데컬처웍스(주)롯데엔터테인먼트   

          등급   장르   감독                                                 배우  \
12  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
13  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
14  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
15  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
16  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   

            일자  개봉 구분     개봉일_dt      일자_dt  
12  2020-06-02  최초 개봉 2020-06-24 2020-06-02  
13  2020-06-03  최초 개봉 2020-06-24 2020-06-03  
14  2020-06-15  최초 개봉 2020-06-24 2020-06-15  
15  2020-06-19  최초 개봉 2020-06-24 2020-06-19  
16  2020-06-23  최초 개봉 2020-06-24 2020-06-23

일자가 개봉일보다 빠른 것은 제거 (개봉전 시사회일 가능성 높음)

In [743]:
temp=temp[temp['개봉일_dt'] <= temp['일자_dt']]

각 영화별로 고유한 첫 개봉일을 '개봉일_대표' 컬럼으로 지정 (날짜 형식 변경한 것을 사용)

In [744]:
temp['개봉일_대표'] = temp.groupby('영화명')['개봉일_dt'].transform('first')

개봉일이 속한 주의 일요일을 '첫주_마감일'로 계산  
월요일 : 0, 일요일 : 6

In [745]:
days_to_sunday = 6 - temp['개봉일_대표'].dt.weekday
temp['첫주_마감일'] = temp['개봉일_대표'] + pd.to_timedelta(days_to_sunday, unit='D')

주차 계산

In [746]:
# 일자_dt가 첫주_마감일보다 작거나 같으면 1일차
is_first_week = temp['일자_dt'] <= temp['첫주_마감일']
# 일자_dt와 첫주_마감일의 차이를 일수로 계산
days_diff = (temp['일자_dt'] - temp['첫주_마감일']).dt.days
# 2주차 부터의 주차 계산
weeks_after_first: int = (days_diff - 1) // 7 + 2 # 값 자체가 주차 숫자
# 첫주차면 1, 아니면 계산된 주차 숫자를 입력
temp['주차'] = np.where(is_first_week, 1, weeks_after_first)

이제 필요없는 컬럼 삭제

In [747]:
temp.drop(columns=['개봉일_dt', '일자_dt', '개봉일_대표', '첫주_마감일'], inplace=True)

확인

In [748]:
temp.tail(5)

영화명         개봉일    매출액  관객수  스크린수  상영횟수 대표국적         배급사       등급  \
682460  힛쳐  2007-08-01  54000    8     2     3   미국  (주)씨지브이신도림  청소년관람불가   
682461  힛쳐  2007-08-01  54000    8     1     2   미국  (주)씨지브이신도림  청소년관람불가   
682462  힛쳐  2007-08-01  26000    4     1     2   미국  (주)씨지브이신도림  청소년관람불가   
682463  힛쳐  2007-08-01  60000    9     1     2   미국  (주)씨지브이신도림  청소년관람불가   
682464  힛쳐  2007-08-01   7000    1     1     1   미국  (주)씨지브이신도림  청소년관람불가   

               장르        감독                         배우          일자  개봉 구분  주차  
682460  액션,범죄,스릴러  데이브 메이어스  숀 빈,소피아 부시,자크리 나이튼,닐 맥도노프  2007-08-08  최초 개봉   2  
682461  액션,범죄,스릴러  데이브 메이어스  숀 빈,소피아 부시,자크리 나이튼,닐 맥도노프  2007-08-10  최초 개봉   2  
682462  액션,범죄,스릴러  데이브 메이어스  숀 빈,소피아 부시,자크리 나이튼,닐 맥도노프  2007-08-11  최초 개봉   2  
682463  액션,범죄,스릴러  데이브 메이어스  숀 빈,소피아 부시,자크리 나이튼,닐 맥도노프  2007-08-12  최초 개봉   2  
682464  액션,범죄,스릴러  데이브 메이어스  숀 빈,소피아 부시,자크리 나이튼,닐 맥도노프  2007-08-13  최초 개봉   3

In [749]:
temp[temp["영화명"]=="#살아있다"].groupby("주차").size()

주차
1     5
2     7
3     7
4     7
5     7
     ..
7     7
8     7
9     5
10    3
19    1
Length: 11, dtype: int64

In [750]:
pd.set_option('display.max_rows', 100)
check_temp = temp.groupby(['영화명', '주차']).size().reset_index(name='counts')
check_temp.head(100)

영화명  주차  counts
0                        #살아있다   1       5
1                        #살아있다   2       7
2                        #살아있다   3       7
3                        #살아있다   4       7
4                        #살아있다   5       7
5                        #살아있다   6       7
6                        #살아있다   7       7
7                        #살아있다   8       7
8                        #살아있다   9       5
9                        #살아있다  10       3
10                       #살아있다  19       1
11                      #아이엠히어   1       4
12                      #아이엠히어   2       7
13                      #아이엠히어   3       7
14                      #아이엠히어   4       7
15                      #아이엠히어   5       2
16                       #위왓치유   1       4
17                       #위왓치유   2       7
18                       #위왓치유   3       7
19                       #위왓치유   4       7
20                       #위왓치유   5       7
21                       #위왓치유   6       3
22                       #위왓치유  26       1
23                       #위왓치유  52       1
24                       #위왓치유  78       1
25    (여자)아이들 월드투어 [아이돌] 인 시네마   1       5
26    (여자)아이들 월드투어 [아이돌] 인 시네마   2       7
27    (여자)아이들 월드투어 [아이돌] 인 시네마   3       2
28                      0.0MHz   1       5
29                      0.0MHz   2       7
30                      0.0MHz   3       7
31                      0.0MHz   4       7
32                      0.0MHz   5       2
33               007 노 타임 투 다이   1       5
34               007 노 타임 투 다이   2       7
35               007 노 타임 투 다이   3       7
36               007 노 타임 투 다이   4       7
37               007 노 타임 투 다이   5       7
38               007 노 타임 투 다이   6       7
39               007 노 타임 투 다이   7       7
40               007 노 타임 투 다이   8       7
41               007 노 타임 투 다이   9       4
42               007 노 타임 투 다이  13       1
43               007 노 타임 투 다이  14       1
44                     007 스펙터   1       5
45                     007 스펙터   2       7
46                     007 스펙터   3       7
47                     007 스펙터   4       7
48                     007 스펙터   5       7
49                     007 스펙터   6       4
50                     007 스펙터   7       1
51                     007 스펙터   8       1
52                  007 카지노 로얄   1       4
53                  007 카지노 로얄   2       7
54                  007 카지노 로얄   3       7
55                  007 카지노 로얄   4       7
56                  007 카지노 로얄   5       7
57                  007 카지노 로얄   6       7
58                  007 카지노 로얄   7       7
59                  007 카지노 로얄   8       3
60                     10 라이브즈   1       4
61                     10 라이브즈   2       7
62                     10 라이브즈   3       7
63                     10 라이브즈   4       7
64                     10 라이브즈   5       7
65                     10 라이브즈   6       7
66                     10 라이브즈   7       4
67                     10 라이브즈  33       2
68                     10 라이브즈  41       2
69                     10 라이브즈  42       1
70                     10 라이브즈  46       1
71                     10 라이브즈  48       1
72                     10 라이브즈  55       1
73                      100 피트   1       4
74                      100 피트   2       7
75                      100 피트   3       2
76                       100미터   1       4
77                       100미터   2       7
78                       100미터   3       2
79                       100미터   4       2
80  100일 동안 100가지로 100퍼센트 행복찾기   1       4
81  100일 동안 100가지로 100퍼센트 행복찾기   2       7
82  100일 동안 100가지로 100퍼센트 행복찾기   3       7
83  100일 동안 100가지로 100퍼센트 행복찾기   4       7
84  100일 동안 100가지로 100퍼센트 행복찾기   5       7
85  100일 동안 100가지로 100퍼센트 행복찾기   6       7
86  100일 동안 100가지로 100퍼센트 행복찾기   7       7
87  100일 동안 100가지로 100퍼센트 행복찾기   8       7
88  100일 동안 100가지로 100퍼센트 행복찾기   9       7
89  100일 동안 100가지로 100퍼센트 행복찾기  10       7
90  100일 동안 100가지로 100퍼센트 행복찾기  11       7
91  100일 동안 100가지로 100퍼센트 행복찾기  12       3
92  100일 동안 100가지로 100퍼센트 행복

In [751]:
pd.set_option('display.max_rows', 10)

각 주차별 일자가 7일이 넘는지 확인 (데이터 중복 검사)

In [752]:
check_temp[check_temp['counts'] > 7]

Empty DataFrame
Columns: [영화명, 주차, counts]
Index: []

문제점  
1. 주차가 확 띄는 경우가 있다 -> 재개봉이 잘 걸러지지 않았다, 따라서 이를 제거  
2. 중간중간 데이터가 비는 경우가 있다 -> 영화별로 얼만큼 비는지 확인

재개봉 주차를 제거  
마지막 개봉주차와의 차이가 1보다 크면 재개봉이라고 판단

In [753]:
check_temp['주차_차이'] = check_temp.groupby('영화명')['주차'].diff()

1~7, 13, 14주차와 같은 경우 14주차는 분명 재개봉이지만 주차 차이는 1이다  
이것은 누적합으로 해결한다

이 방법대로 하면 1~8, 10~주차의 경우 재개봉이 아님에도 판별 못할 수 있다  
그러나 이정도 차이가 나면 결측치 보간을 할때 그 정확도가 매우 떨어진다  
따라서 해당 영화를 아예 제거하는 것보다, 일부분만이라도 사용하는게 나을 수 있지 않을까?  
(어차피 인기있는 영화면 박스오피스 데이터가 누락될 일이 없을 것)

In [754]:
check_temp["재개봉_시작_주차"] = check_temp.groupby('영화명')['주차'].diff()>1 #재개봉 시작 주차 표시
check_temp["재개봉_누적"] = check_temp.groupby("영화명")["재개봉_시작_주차"].cumsum()


In [755]:
pd.set_option('display.max_rows', 1000)
check_temp.head(1000)

영화명    주차  counts  주차_차이  재개봉_시작_주차  재개봉_누적
0                          #살아있다     1       5    NaN      False       0
1                          #살아있다     2       7    1.0      False       0
2                          #살아있다     3       7    1.0      False       0
3                          #살아있다     4       7    1.0      False       0
4                          #살아있다     5       7    1.0      False       0
5                          #살아있다     6       7    1.0      False       0
6                          #살아있다     7       7    1.0      False       0
7                          #살아있다     8       7    1.0      False       0
8                          #살아있다     9       5    1.0      False       0
9                          #살아있다    10       3    1.0      False       0
10                         #살아있다    19       1    9.0       True       1
11                        #아이엠히어     1       4    NaN      False       0
12                        #아이엠히어     2       7    1.0      False       0
13                        #아이엠히어     3       7    1.0      False       0
14                        #아이엠히어     4       7    1.0      False       0
15                        #아이엠히어     5       2    1.0      False       0
16                         #위왓치유     1       4    NaN      False       0
17                         #위왓치유     2       7    1.0      False       0
18                         #위왓치유     3       7    1.0      False       0
19                         #위왓치유     4       7    1.0      False       0
20                         #위왓치유     5       7    1.0      False       0
21                         #위왓치유     6       3    1.0      False       0
22                         #위왓치유    26       1   20.0       True       1
23                         #위왓치유    52       1   26.0       True       2
24                         #위왓치유    78       1   26.0       True       3
25      (여자)아이들 월드투어 [아이돌] 인 시네마     1       5    NaN      False       0
26      (여자)아이들 월드투어 [아이돌] 인 시네마     2       7    1.0      False       0
27      (여자)아이들 월드투어 [아이돌] 인 시네마     3       2    1.0      False       0
28                        0.0MHz     1       5    NaN      False       0
29                        0.0MHz     2       7    1.0      False       0
30                        0.0MHz     3       7    1.0      False       0
31                        0.0MHz     4       7    1.0      False       0
32                        0.0MHz     5       2    1.0      False       0
33                 007 노 타임 투 다이     1       5    NaN      False       0
34                 007 노 타임 투 다이     2       7    1.0      False       0
35                 007 노 타임 투 다이     3       7    1.0      False       0
36                 007 노 타임 투 다이     4       7    1.0      False       0
37                 007 노 타임 투 다이     5       7    1.0      False       0
38                 007 노 타임 투 다이     6       7    1.0      False       0
39                 007 노 타임 투 다이     7       7    1.0      False       0
40                 007 노 타임 투 다이     8       7    1.0      False       0
41                 007 노 타임 투 다이     9       4    1.0      False       0
42                 007 노 타임 투 다이    13       1    4.0       True       1
43                 007 노 타임 투 다이    14       1    1.0      False       1
44                       007 스펙터     1       5    NaN      False       0
45                       007 스펙터     2       7    1.0      False       0
46                       007 스펙터     3       7    1.0      False       0
47                       007 스펙터     4       7    1.0      False       0
48                       007 스펙터     5       7    1.0      False       0
49                       007 스펙터     6       4    1.0      False       0
50                       007 스펙터     7       1    1.0      False       0
51                       007 스펙터     8       1    1.0      False       0
52                    007 카지노 로얄     1       4    NaN      False       0
53                    007 카지노 로얄     2       7    1.0      False       0
54            

In [756]:
pd.set_option('display.max_rows', 10)

In [757]:
check_temp=check_temp[check_temp['재개봉_누적'] == 0]

본 분석은 2주차 까지의 데이터를 사용한다.  
반대로 말하면, 2주차 이후의 데이터가 없으면 해당 영화는 사용할 수 없다

In [758]:
check_temp.shape

(37771, 6)

In [759]:
# '주차' 열에 2가 포함된 영화명을 추출합니다.
movies_with_week_2 = check_temp[check_temp['주차'] == 2]['영화명'].unique()

# 전체 영화명 목록을 가져옵니다.
all_movies = check_temp['영화명'].unique()

# 전체 영화 목록에서 2주차 데이터가 있는 영화 목록을 제외합니다.
movies_without_week_2 = [movie for movie in all_movies if movie not in movies_with_week_2]

# 최종 개수를 출력합니다.
print(len(movies_without_week_2))

492


In [760]:
pd.set_option('display.max_rows', 100)
wk_2 = check_temp[check_temp["영화명"].isin(movies_without_week_2)].groupby(["영화명", "주차"]).size().reset_index(name='counts')
wk_2.head(100)

영화명    주차  counts
0   2024 태민 월드 투어 [이페머럴 게이즈]     1       1
1                      28일 후   826       1
2                      28일 후   827       1
3                      28일 후   828       1
4                      28일 후   829       1
5                       3670     1       1
6                     4인용 식탁   346       1
7                     7인의 황후     1       1
8                       8 마일    40       1
9                       8 마일    41       1
10                 8월의 크리스마스   551       1
11                 8월의 크리스마스   552       1
12                 8월의 크리스마스   553       1
13                 8월의 크리스마스   554       1
14                      K-19   810       1
15                        K2  1278       1
16                   가슴달린 남자  1625       1
17                       가위손  1121       1
18                    가을의 전설  1123       1
19                    감각의 제국   639       1
20                    감각의 제국   640       1
21                       개그맨  1005       1
22                       개그맨  1006       1
23                        걸스     1       1
24                      겟어웨이  2079       1
25                      겟어웨이  2080       1
26                      겟어웨이  2081       1
27                      고래사냥  1261       1
28                      고래사냥  1262       1
29                   고스포드 파크   243       1
30                  고양이를 부탁해   424       1
31                   고양이의 보은    92       1
32                     공각기동대   781       1
33                     공각기동대   782       1
34                     공공의 적   128       1
35                공동경비구역 JSA   198       1
36                        괴테     1       1
37                    굿 윌 헌팅   961       1
38                    굿 윌 헌팅   962       1
39                    굿 윌 헌팅   963       1
40                    굿 윌 헌팅   964       1
41                    굿 윌 헌팅   965       1
42                    굿 윌 헌팅   966       1
43                    굿 윌 헌팅   967       1
44                        그곳     1       1
45                        그날     1       1
46                    그녀의 비밀     1       1
47                       그리스  1631       1
48                    근거리 연애     1       1
49                   글루미 선데이   669       1
50                   글루미 선데이   670       1
51                   글루미 선데이   671       1
52                기쁜 우리 젊은 날  1100       1
53                기쁜 우리 젊은 날  1101       1
54                        기적  1945       1
55                        기적  1946       1
56                  기쿠지로의 여름   302       1
57                       길소뜸  1341       1
58                       길소뜸  1342       1
59                  꼬방동네 사람들  1350       1
60                  꼬방동네 사람들  1351       1
61                        꽃섬   418       1
62                         꿈   922       1
63                         꿈   923       1
64           나도 아내가 있었으면 좋겠다   403       1
65           나도 아내가 있었으면 좋겠다   404       1
66                      나쁜 피   779       1
67                      나쁜 피   780       1
68                      나쁜 피   781       1
69                      나쁜 피   782       1
70                        나츠     1       1
71                        나크   645       1
72                        나크   646       1
73                      남과 여  1674       1
74                      남과 여  1675       1
75                   남자는 괴로워   873       1
76               내 남자친구의 결혼식  1230       1
77                내 책상 위의 천사  1196       1
78                       내츄럴  1652       1
79                      넘버 3   606       1
80                      넘버 3   607       1
81                      넘버 3   608       1
82                      넘버 3   609       1
83                      네이키드  1069       1
84                      네이키드  1070       1
85                      네이키드  1071       1
86                         넬  1237       1
87                        눈물   488       1
88                        눈물   489       1
89                        눈물   490       1
90                        눈물   491       1
91                     뉴욕의 왕  1298       1
92      뉴클래식 프로젝트 미안하다, 사랑

확인 결과 : 최초개봉일과 일자 차이가 많다 -> 박스오피스 계측 전(2003년 11월) 이전의 영화의 관람객 데이터는 없다. 이것이 후에 재개봉하면 자연스레 주차수가 비정상적으로 커진다

In [761]:
check_temp = check_temp[~check_temp["영화명"].isin(movies_without_week_2)]
check_temp.head(20)

영화명  주차  counts  주차_차이  재개봉_시작_주차  재개봉_누적
0    #살아있다   1       5    NaN      False       0
1    #살아있다   2       7    1.0      False       0
2    #살아있다   3       7    1.0      False       0
3    #살아있다   4       7    1.0      False       0
4    #살아있다   5       7    1.0      False       0
5    #살아있다   6       7    1.0      False       0
6    #살아있다   7       7    1.0      False       0
7    #살아있다   8       7    1.0      False       0
8    #살아있다   9       5    1.0      False       0
9    #살아있다  10       3    1.0      False       0
11  #아이엠히어   1       4    NaN      False       0
12  #아이엠히어   2       7    1.0      False       0
13  #아이엠히어   3       7    1.0      False       0
14  #아이엠히어   4       7    1.0      False       0
15  #아이엠히어   5       2    1.0      False       0
16   #위왓치유   1       4    NaN      False       0
17   #위왓치유   2       7    1.0      False       0
18   #위왓치유   3       7    1.0      False       0
19   #위왓치유   4       7    1.0      False       0
20   #위왓치유   5       7    1.0      False       0

In [762]:
pd.set_option('display.max_rows', 10)

# 5. 주차관련 오류 최종 수정

이제 해당 영화명과 주차를 가지고 temp에서 데이터를 최종 필터링한다

In [763]:
result = pd.merge(temp, check_temp[["영화명","주차"]], how="inner")
result.head(10)

영화명         개봉일         매출액     관객수  스크린수  상영횟수 대표국적                배급사  \
0  #살아있다  2020-06-24  1311296000  204066  1662  7223   한국  롯데컬처웍스(주)롯데엔터테인먼트   
1  #살아있다  2020-06-24  1240607900  148991  1637  7311   한국  롯데컬처웍스(주)롯데엔터테인먼트   
2  #살아있다  2020-06-24  1380200100  152550  1713  7839   한국  롯데컬처웍스(주)롯데엔터테인먼트   
3  #살아있다  2020-06-24  2644122800  294255  1882  9221   한국  롯데컬처웍스(주)롯데엔터테인먼트   
4  #살아있다  2020-06-24  2263447540  256133  1869  9044   한국  롯데컬처웍스(주)롯데엔터테인먼트   
5  #살아있다  2020-06-24   556892400   67845  1644  7111   한국  롯데컬처웍스(주)롯데엔터테인먼트   
6  #살아있다  2020-06-24   543848720   67098  1641  7147   한국  롯데컬처웍스(주)롯데엔터테인먼트   
7  #살아있다  2020-06-24   428624500   51659  1441  6137   한국  롯데컬처웍스(주)롯데엔터테인먼트   
8  #살아있다  2020-06-24   368804680   44470  1401  5783   한국  롯데컬처웍스(주)롯데엔터테인먼트   
9  #살아있다  2020-06-24   467173540   51860  1444  6066   한국  롯데컬처웍스(주)롯데엔터테인먼트   

         등급   장르   감독                                                 배우  \
0  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
1  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
2  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
3  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
4  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
5  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
6  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
7  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
8  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
9  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   

           일자  개봉 구분  주차  
0  2020-06-24  최초 개봉   1  
1  2020-06-25  최초 개봉   1  
2  2020-06-26  최초 개봉   1  
3  2020-06-27  최초 개봉   1  
4  2020-06-28  최초 개봉   1  
5  2020-06-29  최초 개봉   2  
6  2020-06-30  최초 개봉   2  
7  2020-07-01  최초 개봉   2  
8  2020-07-02  최초 개봉   2  
9  2020-07-03  최초 개봉   2

In [764]:
pd.set_option('display.max_rows', 100)
result.groupby(["영화명","주차"]).size().reset_index(name='counts').head(100)

영화명  주차  counts
0                        #살아있다   1       5
1                        #살아있다   2       7
2                        #살아있다   3       7
3                        #살아있다   4       7
4                        #살아있다   5       7
5                        #살아있다   6       7
6                        #살아있다   7       7
7                        #살아있다   8       7
8                        #살아있다   9       5
9                        #살아있다  10       3
10                      #아이엠히어   1       4
11                      #아이엠히어   2       7
12                      #아이엠히어   3       7
13                      #아이엠히어   4       7
14                      #아이엠히어   5       2
15                       #위왓치유   1       4
16                       #위왓치유   2       7
17                       #위왓치유   3       7
18                       #위왓치유   4       7
19                       #위왓치유   5       7
20                       #위왓치유   6       3
21    (여자)아이들 월드투어 [아이돌] 인 시네마   1       5
22    (여자)아이들 월드투어 [아이돌] 인 시네마   2       7
23    (여자)아이들 월드투어 [아이돌] 인 시네마   3       2
24                      0.0MHz   1       5
25                      0.0MHz   2       7
26                      0.0MHz   3       7
27                      0.0MHz   4       7
28                      0.0MHz   5       2
29               007 노 타임 투 다이   1       5
30               007 노 타임 투 다이   2       7
31               007 노 타임 투 다이   3       7
32               007 노 타임 투 다이   4       7
33               007 노 타임 투 다이   5       7
34               007 노 타임 투 다이   6       7
35               007 노 타임 투 다이   7       7
36               007 노 타임 투 다이   8       7
37               007 노 타임 투 다이   9       4
38                     007 스펙터   1       5
39                     007 스펙터   2       7
40                     007 스펙터   3       7
41                     007 스펙터   4       7
42                     007 스펙터   5       7
43                     007 스펙터   6       4
44                     007 스펙터   7       1
45                     007 스펙터   8       1
46                  007 카지노 로얄   1       4
47                  007 카지노 로얄   2       7
48                  007 카지노 로얄   3       7
49                  007 카지노 로얄   4       7
50                  007 카지노 로얄   5       7
51                  007 카지노 로얄   6       7
52                  007 카지노 로얄   7       7
53                  007 카지노 로얄   8       3
54                     10 라이브즈   1       4
55                     10 라이브즈   2       7
56                     10 라이브즈   3       7
57                     10 라이브즈   4       7
58                     10 라이브즈   5       7
59                     10 라이브즈   6       7
60                     10 라이브즈   7       4
61                      100 피트   1       4
62                      100 피트   2       7
63                      100 피트   3       2
64                       100미터   1       4
65                       100미터   2       7
66                       100미터   3       2
67                       100미터   4       2
68  100일 동안 100가지로 100퍼센트 행복찾기   1       4
69  100일 동안 100가지로 100퍼센트 행복찾기   2       7
70  100일 동안 100가지로 100퍼센트 행복찾기   3       7
71  100일 동안 100가지로 100퍼센트 행복찾기   4       7
72  100일 동안 100가지로 100퍼센트 행복찾기   5       7
73  100일 동안 100가지로 100퍼센트 행복찾기   6       7
74  100일 동안 100가지로 100퍼센트 행복찾기   7       7
75  100일 동안 100가지로 100퍼센트 행복찾기   8       7
76  100일 동안 100가지로 100퍼센트 행복찾기   9       7
77  100일 동안 100가지로 100퍼센트 행복찾기  10       7
78  100일 동안 100가지로 100퍼센트 행복찾기  11       7
79  100일 동안 100가지로 100퍼센트 행복찾기  12       3
80                         10년   1       4
81                         10년   2       7
82                         10년   3       4
83                         10년   4       1
84                         10분   1       4
85                         10분   2       7
86                         10분   3       7
87                         10분   4       7
88                         10분   5       4
89                         10분   6       3
90                         10분   7       1
91                         10분   8       1
92                         1

In [765]:
pd.set_option('display.max_rows', 10)

In [766]:
result["영화명"].nunique()

4773

최종 4773개의 영화가 추출되었다

# 6. 공휴일 등 날짜 결합

공휴일과 날짜가 있는 데이터를 일자를 기준으로 결합한다

In [767]:
date_data = pd.read_csv("../calendar_dataset.csv", encoding='utf-8-sig')
date_data.head()

Date  Holiday  Moonhwa
0  2003-01-01        1        0
1  2003-01-02        0        0
2  2003-01-03        0        0
3  2003-01-04        1        0
4  2003-01-05        1        0

In [768]:
date_data.isna().sum() 

Date       0
Holiday    0
Moonhwa    0
dtype: int64

In [769]:
date_data["Date"] = pd.to_datetime(date_data["Date"], errors='coerce')
date_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8340 entries, 0 to 8339
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   Date     8340 non-null   datetime64[ns]
 1   Holiday  8340 non-null   int64         
 2   Moonhwa  8340 non-null   int64         
dtypes: datetime64[ns](1), int64(2)
memory usage: 195.6 KB


In [770]:
result["일자"] = pd.to_datetime(result["일자"], errors='coerce')
result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197697 entries, 0 to 197696
Data columns (total 15 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   영화명     197697 non-null  object        
 1   개봉일     197697 non-null  object        
 2   매출액     197697 non-null  int64         
 3   관객수     197697 non-null  int64         
 4   스크린수    197697 non-null  int64         
 5   상영횟수    197697 non-null  int64         
 6   대표국적    197697 non-null  object        
 7   배급사     197322 non-null  object        
 8   등급      197631 non-null  object        
 9   장르      197578 non-null  object        
 10  감독      191135 non-null  object        
 11  배우      184300 non-null  object        
 12  일자      197697 non-null  datetime64[ns]
 13  개봉 구분   197697 non-null  object        
 14  주차      197697 non-null  int64         
dtypes: datetime64[ns](1), int64(5), object(9)
memory usage: 22.6+ MB


일자 기준으로 병합

In [771]:
result.shape

(197697, 15)

In [772]:
temp_result = pd.merge(result, date_data, left_on = "일자", right_on = "Date", how = "left")
temp_result.shape


(197697, 18)

In [773]:
temp_result.drop(columns=["일자"], inplace=True)
temp_result.head(2)

영화명         개봉일         매출액     관객수  스크린수  상영횟수 대표국적                배급사  \
0  #살아있다  2020-06-24  1311296000  204066  1662  7223   한국  롯데컬처웍스(주)롯데엔터테인먼트   
1  #살아있다  2020-06-24  1240607900  148991  1637  7311   한국  롯데컬처웍스(주)롯데엔터테인먼트   

         등급   장르   감독                                                 배우  \
0  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   
1  15세이상관람가  드라마  조일형  유아인,박신혜,전배수,고나영,권용채,김경태,김다영,김단비,김라희,김미래,김미래,김윤...   

   개봉 구분  주차       Date  Holiday  Moonhwa  
0  최초 개봉   1 2020-06-24        0        1  
1  최초 개봉   1 2020-06-25        0        0

`펜데믹 구분` 은 아래와 같은 규칙으로 생성한다

1. 2020년 1월 까지는 `0`
2. 2020년 2월부터 2023년 4월까지는 `1`
3. 2023년 5월부터는 `2`

In [774]:
def get_pandemic_flag(date):
    if date < pd.Timestamp('2020-02-01'):
        return 0
    elif date < pd.Timestamp('2023-05-01'):
        return 1
    else:
        return 2

temp_result['Pandemic'] = temp_result['Date'].apply(get_pandemic_flag)

# 7. 최종 사용 컬럼 및 영문명 변경

In [777]:
final_result = \
temp_result.rename(columns={
    "영화명": "Movie_Title",
    "개봉일" : "Release_Date",
    "관객수": "Audience_Count",
    "상영횟수":"Show_Count",
    "대표국적":"Main_Country",
    "배급사":"Distributor",
    "등급":"Grade",
    "장르":"Genre",
    "주차":"Week"
    })

In [779]:
final_result.drop(columns=["매출액", "스크린수", "감독", "배우", "개봉 구분"], inplace=True)

In [780]:
final_result.to_csv("boxoffice_data.csv", index=False, encoding='utf-8-sig')